### DISCUSSION TITLE
How to have Status "Spare" in the Devices Page and Status "Store" in the Location Page

### DISCUSSION THREAD

**Discussion by mahmoudkhalil2** [netbox-community/netbox#12240] [Q&A]
Hello, I need to add some devices but as Spare status and I didn't find spare status in the device status tab, so how to add device with spare status?
also, how to add Location in the location page with status "Store" because this is a Store room?

**Comment by mtinberg** [comment]
While you might be able to add custom statuses in the configuration, the way we track dedicated hotspares is to set the Device Role to "hotspare" with the status of Active, Inventory, etc. and for Devices or Modules that have another role such as QA/Lab that also serve as hotspares (for high-value items) I use a Tag of "hotspare" to mark them.  That works well for us.  We use custom_fields on Device Type and Module Type to record our expected number of hotspares and a Custom Report to email if there are Device/Module Types with Device/Module records that don't have sparing configured or if the actual number of configured hotspares doesn't match. Additionally, we configure custom_fields with desired coldspare levels which the inventory team pulls via API to update the inventory management system and min/max for their cycle counts

As far as Location I would just use the Name and Description to indicate which room and what it's for and not the Status field.  Locations can also be hierarchal if you need to create sub-locations in a room.  So you might have a Site of "Main Office" and a Location of room "1234" with a Description of "IT Store Room" and sub-locations of "Shelf 1" or something.  That said Netbox doesn't excel as an inventory tracking system itself, for tracking POs, on-shelf inventory, check-out/check-in, equipment lifecycle, etc. as it's focus is storing config management data for active devices, so setting up an IT asset-tracking tool like Snipe-IT and writing some API integration for hand-off/audit between tools makes sense.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: mahmoudkhalil2 ***@***.***>
Sent: Thursday, April 13, 2023 8:27 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] How to have Status "Spare" in the Devices Page and Status "Store" in the Location Page (Discussion #12240)


Hello, I need to add some devices but as Spare status and I didn't find spare status in the device status tab, so how to add device with spare status?
also, how to add Location in the location page with status "Store" because this is a Store room?

**Reply by jackpickleweed** [reply]
To create a custom device status, edit /opt/netbox/netbox/dcim/choices.py and find "class DeviceStatusChoices(ChoiceSet):" and enter your custom value. Example for "Spare" below: 

class DeviceStatusChoices(ChoiceSet):
	key = 'Device.status'
	STATUS_OFFLINE = 'offline'
	STATUS_ACTIVE = 'active'
	STATUS_PLANNED = 'planned'
	STATUS_STAGED = 'staged'
	STATUS_FAILED = 'failed'
	STATUS_INVENTORY = 'inventory'
	STATUS_DECOMMISSIONING = 'decommissioning'
	STATUS_Spare = 'spare'
	
	CHOICES = [  (STATUS_OFFLINE, _('Offline'), 'gray'),
	(STATUS_ACTIVE, _('Active'), 'green'),
	(STATUS_PLANNED, _('Planned'), 'cyan'),
	(STATUS_STAGED, _('Staged'), 'blue'),
	(STATUS_FAILED, _('Failed'), 'red'),
	(STATUS_INVENTORY, _('Inventory'), 'purple'),
	(STATUS_DECOMMISSIONING, _('Decommissioning'), 'yellow'),
	(STATUS_Spare, _('Spare'), 'black'),                      ]             
	
Remember if you upgrade Netbox, make sure you copy this back in.

**Reply by mtinberg** [reply]
I don't think that's the preferred way to modify and customize a status, as you note you'd have to port this change forward on every upgrade, adding FIELD_CHOICES to the configuration is the built-in mechanism to make this kind of customization that will survive upgrades

https://demo.netbox.dev/static/docs/configuration/data-validation/

**Reply by jackpickleweed** [reply]
You're absolutely correct, I missed that in the documentation. Thank you. 

The correct way to do this is edit: 

/opt/netbox/netbox/netbox/configuration.py 

And add the code below to the top of the configuration: 

```
# Custom settings

# Appends the predefined device status choices with an additional "Spare" selection, color coded as black                                                                                                                                                                                                                                                                                                                                                                                                               
FIELD_CHOICES = {                                                                                                                                                                   
       'dcim.Device.status+': (                                                                                                                                                                                                                       
              ('spare', 'Spare', 'black'),                                                                                                                                                                                                       
       )                                                                                                                                                                                                                                  } 

```
